### PR TITLE
Fix nursery crashing during a guest run when the host loop calls `start_soon(needs_sniffio)`

### LINKED ISSUE #2183: trio's guest mode breaks sniffio.current_async_library in certain cases
I am opening this bug as requested on Gitter (see link below).

## Summary
When running trio as a guest (in a `tkinter` event loop in this case), and using `trio.Nursery::start_soon()` from the host event loop to schedule work to be done in the trio event loop, `sniffio.current_async_library()` doesn't detect the trio loop and raises.

## Details
When running trio in guest mode in the `tkinter` event loop, the docs advise to have both loop have the same lifetime. This is achieved by passing a `done_callback` to `trio.start_guest_run()` that terminates the host loop. The closing mechanism is then to terminate the trio loop (most likely by cancelling it), which in turn terminates the host loop.

The doc does not provide examples on how to start async function inside the trio loop from the host loop. When reading it, it was unclear to me what functions or methods are safe to call from *outside* the trio loop. I considered `trio.Nursery::start_soon`, `trio.SendChannel::put_nowait()`, `trio.lowlevel.TrioToken::run_sync_soon()` and `trio.lowlevel.spawn_system_task()`.

As nurseries seemed like an excellent control flow mechanism and a central concept in trio, I tried the Nursery. A Nursery is created in the trio loop, immediately calls `nursery.start_soon(trio.sleep_forever)` to remain available, and the host loop gets a reference to this nursery to schedule async function in it.

This works for trio code, but seem to breaks `sniffio.current_async_library()`, which raises an exception. I my particular use-case, `sniffio` was used under the hood by `asks`, which was used for http requests.

I wrote a (relatively) small example code to showcase the issue, as well as the fix proposed on Gitter. See the link below.

I am unsure if this is an actual bug in sniffio, or trio guest mode, or bad code on my part.
Maybe the doc of `trio.lowlevel.start_guest_run()` could be improved to show how to do this correctly (e.g. using memory channels), so as to avoid future confusion or unsafe code.

## Additional resources
- discussion on gitter https://gitter.im/python-trio/general?at=619f8ac5c6149e53494c05b7
- reproducing code: https://gist.github.com/QuentinSoubeyran/138f0784f9d596f5727ca2bea830ffcd

### Versions:
- OS: Ubuntu 20.04
- python 3.10
- trio 0.19.0
- sniffio 1.2.0

If any info is missing, I'll gladly improve this issue !

## Words of thank
Thank you all for the trio lib, it looks awesome ! I was extremely glade to find about guest mode, it really feels like the *correct* way to use async I/O and a UI together.

### PR DESCRIPTION
**Pull request by gschaffner**
Closes #2183.

### CODE REVIEW

**Comment by richardsheridan** [comment]
The new commit I pushed merges #2574 into this PR, in essentially the state that #2574 would be by applying your review comments to it. Using this roundabout method has the advantage of acknowledging your contribution and inviting you to join the python-trio org!  If you acknowledge that you are ok with this, I'll bring it all up to date with master and merge.

**Comment by richardsheridan** [comment]
I also added a test for your edge case of a function returning a coroutine!

File: newsfragments/2574.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix :func:`sniffio.current_async_library` when Trio tasks are spawned from a non-Trio context (such as when using trio-asyncio). Previously, a regular Trio task would inherit the non-Trio library name, and spawning a system task would cause the non-Trio caller to start thinking it was Trio.
```
[the remark this answers is not in the document]
Comment (richardsheridan) [reply newsfragments/2574.bugfix.rst]: Oops! I think the commit message phrasing fits better if the link points here to the fix. I'll push a clean renumber commit when CI is working again.
